web3.js/src/connection.ts

```diff
@@ -2945,9 +2959,46 @@ export class Connection {
           () => resolve({__type: TransactionStatus.TIMED_OUT, timeoutMs}),
           timeoutMs,
         );
+      } else if (Object.prototype.hasOwnProperty.call(strategy, 'nonceHash')) {
+        const {nonceAccount, nonceHash} =
+          strategy as NonceBasedTransactionConfirmationStrategy;
+        let currentNonceHash = nonceHash;
+        const getCurrentNonceHash = async () => {
+          try {
+            const nonceAccountData = await this.getNonce(
+              nonceAccount,
+              commitment,
+            );
+            return nonceAccountData!.nonce;
+          } catch (_e) {
+            // If for whatever reason we can't reach/read the nonce
+            // account, just keep using the last-known value.
+            return currentNonceHash;
+          }
+        };
+        (async () => {
+          currentNonceHash = await getCurrentNonceHash();
+          if (done) return;
+          while (nonceHash === currentNonceHash) {
+            await sleep(2000);
+            if (done) return;
+            currentNonceHash = await getCurrentNonceHash();
+            if (done) return;
+          }
+          if (done) return;
+          resolve({__type: TransactionStatus.NONCE_INVALID});
```

Comment:
This could race with signature confirmation right? What if this polling logic detects that the nonce value was advanced before the signature subscription is notified? This could happen if the subscription uses a different RPC server than the server responding to the `getNonce` request. Maybe we can check the signature explicitly (and make use of  `minContextSlot`) before returning an error

Reply:
Oh, interesting. Indeed it could.

This is everything that's bad about the RPC model of client communication. I know that there's this appetite to push logic _out_ of the RPC, but race conditions are the consequence. If I could push all of the transaction confirmation logic _down_ into the RPC (nonce, nonce account, context slot, signature string, the whole thing) then one endpoint could manage the whole thing and respond or throw over a single connection; no races.